9-react.ts+TDD/src/App.test.tsx

```diff
@@ -0,0 +1,85 @@
+import faker from "faker";
+import puppeteer, { ElementHandle, Page, Browser } from "puppeteer";
+import { TodoItemDTO } from "../../common/interfaces";
+
+
+const todoItem: TodoItemDTO = {
+    text: faker.random.words(3),
+    id: faker.datatype.uuid(),
+    isChecked: faker.datatype.boolean(),
+  };
+
+
+const initTestEnv = async (): Promise<Page> => {
+  const browser: Browser = await puppeteer.launch({
+    headless: true,
+    devtools: false,
+  });
+
+  const page: Page = await browser.newPage();
+
+  page.emulate({
+    viewport: {
+      width: 500,
+      height: 900
+    },
+    userAgent: ''
+  });
+
+  await page.goto('http://localhost:8080/');
+  await page.waitForSelector('.appBodyContainer-0-2-3');
+
+  return page;
+}
+
+
+const addNewItem = async (page: Page): Promise<void> => {
+  await page.click(".newItemButton-0-2-8");
+  await page.type("input[name=add-new-item]", todoItem.text);
+  await page.click(".submitButton-0-2-12");
+}
+
+describe('E2E', () => {
+    test('Add new item', async () => {
+      const page: Page = await initTestEnv();
+      await addNewItem(page);
+      await page.waitForSelector('div[data-hook=todo-item]');
```

Comment:
1. have a line break between the `given, when then` sections.
2. your test checks for the existence fo the div but it doesn't checks its content
3. you should use drivers for e2e tests as well. I don't want to see `waitForSelector` in the test file itself.. i want to see something like `driver.get.firstTodo()`

Reply:
the reasons for building the drivers
1. readability
2. code dup (aka. not writing 20 times `waitForSelector('div[data-hook=todo-item]')`)
3. less important reason but it can be an abstraction for the implementation.. i.e in case of a huge re-write you can run the exact same test suite just with a new driver implementation (b/c the core functionality shouldn't change)